### DISCUSSION TITLE
Converting archr objects to seurat or singlecellexperiment objects

### DISCUSSION THREAD

**Discussion by sel041** [GreenleafLab/ArchR#871] [Questions / Documentation] [answered]
Hi archr team, 

I was wondering if I can convert archr objects to seurat or singlecellexperiment objects. 

Thanks!

**Comment by rcorces** [comment, marked as answer]
This is not currently possible. As you can imagine, the architecture of ArchR and Seurat are not super compatible. You can export certain matrices from ArchR using `getMatrixFromProject()` and this should allow you do get the data you want out.

**Reply by sel041** [reply]
Got it. I will use getMatrixFromProject function. Thank you so much for your prompt answer!

**Reply by sylestiel** [reply]
@synlee

How did you go from GetMatrixFromProject to Seurat Object?

**Reply by dagarfield** [reply]
The resulting object can be converted directly into a `SingleCellExperiment` and then into Seurat from there.

**Reply by pengxin2019** [reply]
Hi dagarfield,
Thanks for providing your tips here. I am trying to exporting the gene score matrix using getMatrixFromProject
![Screen Shot 2022-03-21 at 9 14 12 AM](https://user-images.githubusercontent.com/54915282/159279756-a3e52684-8ccc-432d-9099-2d4045af36ed.png)


I am not too familiar with ArchR data. Can you give me more hint on how to convert this so called GeneScoreMatrix to SingleCellExperiment? I am quite confused since the printed gene.score.mtx is simply a character which is not a high dimensional matrix as I thought.

Thanks in advance

Penny

> The resulting object can be converted directly into a `SingleCellExperiment` and then into Seurat from there.

**Reply by rcorces** [reply]
The correct function is `getMatrixFromProject()`, not `getAvailableMatrices()`. Please see the documentation for further info

**Reply by pengxin2019** [reply]
> The correct function is `getMatrixFromProject()`, not `getAvailableMatrices()`. Please see the documentation for further info

Thanks Ryan for the prompt response!!

**Reply by sylestiel** [reply]
@dagarfield 

Hi,

 i did the followinging,

proj <- getMatrixFromProject(proj, useMatrix = "GeneExpressionMatrix")
proj <- as(proj, "SingleCellExperiment")

It works upto the sce step.

Now for converting this to a Seurat object.

I'm stumped!

The project i'm working with is a subset of a larger project.

Suggestions?

**Reply by dagarfield** [reply]
See here: https://satijalab.org/seurat/archive/v3.1/conversion_vignette.html
Or just keep it as a SCE and do your analysis there: https://bioconductor.org/books/release/OSCA/

**Reply by sylestiel** [reply]
Thanks!!

**Reply by Ping-lin14** [reply]
@sylestiel 

Hi! I wonder if you have successfully converted the Archr object to the Seurat object?
I tried it and I didn't succeed.
Here is my code:
```
proj_gs <- getMatrixFromProject(proj, useMatrix = "GeneScoreMatrix")
proj_sce <- as(proj_gs, "SingleCellExperiment")
library(scater)
proj_sce <- runPCA(proj_sce)
```
```
Error in assay(x, exprs_values) : 
  'assay(<SingleCellExperiment>, i="character", ...)' invalid subscript 'i'
'logcounts' not in names(assays(<SingleCellExperiment>))
```

**Reply by dagarfield** [reply]
@Ping-lin14 
I'm not seeing the Seurat conversion step in your code. Possible you missed a step? In any case, you'll need to do some form of normalization before PCA, though this is outside the scope of the ArchR git.

**Reply by Ping-lin14** [reply]
@dagarfield 
I think I didn't express myself clearly.
I followed [https://satijalab.org/seurat/articles/conversion_vignette.html](https://satijalab.org/seurat/articles/conversion_vignette.html).
According to the demo, `runPCA` needs to be executed before executing `as.Seurat`.
However, I failed, I am not sure if there is one step missing.

**Comment by dagarfield** [comment]
1) runPCA doesn't (I don't think) need to be run before converstion
2) In any case, for both conversion and runPCA, you need logcounts.

Try this:

```
counts <- assay(proj_sce, "counts")
libsizes <- colSums(counts)
size.factors <- libsizes/mean(libsizes)
logcounts(proj_sce) <- log2(t(t(counts)/size.factors) + 1)
```

**Reply by pengxin2019** [reply]
Hi there,
I just tried the solution you mentioned:

```
proj_gs <- getMatrixFromProject(proj.doublet.filtered.harmony, useMatrix = "GeneScoreMatrix")
proj_gs

# class: SummarizedExperiment 
# dim: 30311 19265 
# metadata(0):
#   assays(1): GeneScoreMatrix
# rownames: NULL
# rowData names(6): seqnames start ... name idx
# colnames(19265): pbmc.6800.2x#GGAGAACAGGTTCGAG-1
# pbmc.6800.2x#GAACGTTCAATGTGCT-1 ... pbmc.6800.1x#GATGATCCACTCGTGG-1
# pbmc.6800.1x#TACGGATCAGGACCTC-1
# colData names(14): DoubletEnrichment DoubletScore ... TSSEnrichment
# projColData[, colnames(projColData) %ni% colnames(colData)]


BiocManager::install("SingleCellExperiment")
library(SingleCellExperiment)
proj_sce <- as(proj_gs, "SingleCellExperiment")
# class: SingleCellExperiment 
# dim: 30311 19265 
# metadata(0):
#   assays(1): GeneScoreMatrix
# rownames: NULL
# rowData names(6): seqnames start ... name idx
# colnames(19265): pbmc.6800.2x#GGAGAACAGGTTCGAG-1
# pbmc.6800.2x#GAACGTTCAATGTGCT-1 ... pbmc.6800.1x#GATGATCCACTCGTGG-1
# pbmc.6800.1x#TACGGATCAGGACCTC-1
# colData names(14): DoubletEnrichment DoubletScore ... TSSEnrichment
# projColData[, colnames(projColData) %ni% colnames(colData)]
# reducedDimNames(0):
#   altExpNames(0):


> proj_sce@assays@data
List of length 1
names(1): GeneScoreMatrix
```
As you can see from above, there is not `counts`assay. So, I am using `GeneScoreMatrix`as `counts` below:

```
counts <- assay(proj_sce, "GeneScoreMatrix")
# 30311 19265
counts(proj_sce) <- counts

libsizes <- colSums(counts)
size.factors <- libsizes/mean(libsizes)
logcounts(proj_sce) <- log2(t(t(counts)/size.factors) + 1)
# > proj_sce


```
However, its complaining that there is no rownames in the matrix when I ran `as.Seurat`:
```
library(Seurat)
#proj_sce <- runPCA(proj_sce)
proj_sce <- as.Seurat(proj_sce, counts = "counts", data = "logcounts")#

No feature names (rownames) names present in the input matrix
```
I have checked the arguments of `getMatrixFromProject` and it seems like there is no argument there enables the rownames to the 
`GeneScoreMatrix` assay. Do you have any suggestions?

Thanks
Best

**Comment by dagarfield** [comment]
Add row names. ;)
rownames(proj_sce)<-"whatever_you_want_your_rownames_to_be"
